src/main/java/frc/robot/subsystems/swerve/ModuleIOReal.java

```diff
@@ -202,17 +200,22 @@ public void updateInputs(ModuleIOInputs inputs) {
     inputs.turnAppliedVolts = turnAppliedVolts.getValueAsDouble();
     inputs.turnCurrentAmps = new double[] {turnCurrent.getValueAsDouble()};
 
-    inputs.odometryTimestamps =
-        timestampQueue.stream().mapToDouble((Double value) -> value).toArray();
+    var samples =
+        PhoenixOdometryThread.getInstance()
+            .samplesSince(lastUpdate, ImmutableSet.of(drivePosition, turnPosition));
+    if (!samples.isEmpty()) {
+      lastUpdate = samples.get(samples.size() - 1).timestamp();
+    }
+
+    inputs.odometryTimestamps = samples.stream().mapToDouble(s -> s.timestamp()).toArray();
     inputs.odometryDrivePositionsMeters =
-        drivePositionQueue.stream().mapToDouble((Double value) -> value).toArray();
+        samples.stream().filter(s -> s != null).mapToDouble(s -> s.values().get(drivePosition)).toArray();
     inputs.odometryTurnPositions =
-        turnPositionQueue.stream()
-            .map(Rotation2d::fromRotations) // should be after offset + gear ratio
+        samples.stream()
+            // should be after offset + gear ratio
+            .filter(s -> s != null)
```

Comment:
@Lewis-Seiden This (and the other filter) aren't quite right. Samples should never be null. The thing we're protecting from is that the call to .get might ask for a signal that wasn't collected. So we need to break the map into "map to the result of the get call, filter out the nulls, then map into the rotations".

Does that make sense?